File: sdk/go/configstore/manager.go
package configstore

import (
	"context"
	"encoding/json/jsontext"
	"encoding/json/v2"
	"errors"
	"fmt"
	"maps"
	"strings"
	"sync"

	"github.com/Suhaibinator/kms/sdk/go/kmsclient"
)

// Manager owns managed release policy and the background lower-level loader.
type Manager struct {
	loader  *kmsclient.ReleaseLoader
	options Options
	prepare PrepareFunc

	readyCh   chan struct{}
	done      chan struct{}
	readyOnce sync.Once
	reportMu  sync.Mutex

	mu              sync.RWMutex
	ready           bool
	observed        ReleaseIdentity
	applied         ReleaseIdentity
	divergent       bool
	lastReportedKey string
	lastRejectedKey string
	waitErr         error
}

// String prevents formatting from traversing the retained lower-level loader,
// which owns the private binding-key map for the manager's lifetime.
func (m *Manager) String() string {
	if m == nil {
		return "Manager<nil>"
	}
	return fmt.Sprintf("Manager{release=%q}", m.options.Release)
}

func (m *Manager) GoString() string { return m.String() }

func (m *Manager) Format(f fmt.State, verb rune) {
	if verb == 'q' {
		_, _ = fmt.Fprintf(f, "%q", m.String())
		return
	}
	_, _ = fmt.Fprint(f, m.String())
}

type managerJSON struct {
	Release string `json:"release"`
}

func (m *Manager) safeProjection() *managerJSON {
	if m == nil {
		return nil
	}
	return &managerJSON{Release: m.options.Release}
}

func (m *Manager) MarshalJSON() ([]byte, error) {
	return json.Marshal(m.safeProjection())
}

func (m *Manager) MarshalJSONTo(out *jsontext.Encoder) error {
	return json.MarshalEncode(out, m.safeProjection())
}

// Start validates its generated contract, starts ReleaseLoader in the
// background, and waits until the initial generation has been atomically
// published. No Manager is returned before generated Current is usable.
func Start(
	ctx context.Context,
	client *kmsclient.Client,
	options Options,
	prepare PrepareFunc,
) (*Manager, error) {
	if ctx == nil {
		return nil, errors.New("configstore: Start requires a context")
	}
	if client == nil {
		return nil, errors.New("configstore: Start requires a kmsclient client")
	}
	if prepare == nil {
		return nil, errors.New("configstore: Start requires a prepare function")
	}
	options.Release = strings.TrimSpace(options.Release)
	if options.Release == "" {
		return nil, errors.New("configstore: Options.Release is required")
	}
	if options.OnDefaultMismatch == nil {
		return nil, errors.New("configstore: Options.OnDefaultMismatch is required")
	}
	if err := validateContract(options.Contract); err != nil {
		return nil, err
	}
	options.Contract = append([]ContractEntry(nil), options.Contract...)
	bindingKeys := maps.Clone(options.BindingKeys)
	// The manager retains Options for callbacks and status bookkeeping; keep
	// credentials only in the lower-level loader.
	options.BindingKeys = nil

	manager := &Manager{
		options: options,
		prepare: prepare,
		readyCh: make(chan struct{}),
		done:    make(chan struct{}),
	}
	validateManifest := manifestValidator(options.Contract)
	loader, err := kmsclient.NewReleaseLoader(client, kmsclient.ReleaseLoaderConfig{
		Name:                options.Release,
		ReconcileInterval:   options.ReconcileInterval,
		SecretTokenProvider: options.SecretTokenProvider,
		BindingKeys:         bindingKeys,
		ValidateManifest: func(ctx context.Context, manifest kmsclient.ReleaseManifest) error {
			identity := releaseIdentityFromManifest(manifest)
			manager.mu.Lock()
			manager.observed = identity
			manager.mu.Unlock()
			err := validateManifest(ctx, manifest)
			if err != nil {
				manager.notifyCandidateRejected(identity, err)
			}
			return err
		},
		MaxConcurrentFetches: options.MaxConcurrentFetches,
		InstanceID:           options.InstanceID,
	})
	if err != nil {
		return nil, err
	}

	manager.loader = loader
	go manager.run(ctx)

	select {
	case <-manager.readyCh:
		return manager, nil
	case <-manager.done:
		manager.mu.RLock()
		ready := manager.ready
		waitErr := manager.waitErr
		manager.mu.RUnlock()
		if ready {
			return manager, nil
		}
		if waitErr == nil {
			waitErr = errors.New("configstore: release loader stopped before initial publication")
		}
		return nil, waitErr
	}
}

func (m *Manager) run(ctx context.Context) {
	err := m.loader.Run(ctx, m.prepareSnapshot)
	m.mu.Lock()
	if m.ready && ctx.Err() != nil && isContextError(err) {
		err = nil
	}
	m.waitErr = err
	m.mu.Unlock()
	close(m.done)
}

func isContextError(err error) bool {
	return errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded)
}

func (m *Manager) prepareSnapshot(
	ctx context.Context,
	snapshot kmsclient.ReleaseSnapshot,
) (kmsclient.PreparedRelease, error) {
	return m.prepareWithIdentity(ctx, snapshot, ReleaseIdentityFromSnapshot(snapshot))
}

func (m *Manager) prepareWithIdentity(
	ctx context.Context,
	snapshot kmsclient.ReleaseSnapshot,
	identity ReleaseIdentity,
) (kmsclient.PreparedRelease, error) {
	m.mu.Lock()
	m.observed = identity
	startup := !m.ready
	m.mu.Unlock()

	candidate, err := m.prepare(ctx, snapshot)
	if err != nil {
		newAbort(candidate.Abort)()
		m.notifyCandidateRejected(identity, err)
		return nil, err
	}
	abort := newAbort(candidate.Abort)
	if candidate.Publish == nil {
		abort()
		err := Reject(RejectConfigValidationFailed,
			errors.New("configstore: prepared candidate Publish is required"))
		m.notifyCandidateRejected(identity, err)
		return nil, err
	}
	if err := ctx.Err(); err != nil {
		abort()
		return nil, err
	}

	differences := cloneDifferences(candidate.DefaultDifferences)
	restartFields := append([]string(nil), candidate.RestartRequiredFields...)
	if !startup && len(restartFields) != 0 {
		abort()
		err := rejectWithPaths(RejectRestartRequired, &restartRequiredError{fields: restartFields}, restartFields)
		m.notifyCandidateRejected(identity, err)
		return nil, err
	}

	divergent := len(differences) != 0
	if divergent {
		// Divergence from source defaults is reported, never refused: a process
		// must be able to restart onto whatever release is active. The report is
		// the operator's signal to reconcile code and KMS.
		phase := PhaseRuntime
		if startup {
			phase = PhaseStartup
		}
		m.reportOnce(identity, newDefaultMismatchReport(phase, MismatchError, identity, differences))
	}

	var changes []FieldChange
	if !startup {
		changes = cloneChanges(candidate.Changed)
	}
	return &managedPrepared{
		manager:    m,
		publish:    candidate.Publish,
		abort:      abort,
		identity:   identity,
		divergent:  divergent,
		fieldCount: len(differences),
		startup:    startup,
		changes:    changes,
		groups:     candidate.Groups,
	}, nil
}

func (m *Manager) notifyCandidateRejected(identity ReleaseIdentity, err error) {
	var candidateErr *CandidateError
	if !errors.As(err, &candidateErr) {
		return
	}
	callback := m.options.OnCandidateRejected
	if callback == nil {
		return
	}
	category := RejectionCategory(candidateErr.ReleaseRejectionCategory())
	key := identity.dedupeKey()
	m.mu.Lock()
	if m.lastRejectedKey == key {
		m.mu.Unlock()
		return
	}
	// Record before invoking application code. A panic must neither alter
	// candidate admission nor turn reconciliation into a callback storm.
	m.lastRejectedKey = key
	m.mu.Unlock()

	report := newCandidateRejectionReport(category, identity, candidateErr.pathsCopy())
	func() {
		defer func() { _ = recover() }()
		callback(report)
	}()
}

// reportOnce delivers a mismatch report at most once per release identity.
// The callback is an observer: a panic is swallowed and never influences
// candidate admission, so a broken logger cannot refuse a restart.
func (m *Manager) reportOnce(identity ReleaseIdentity, report DefaultMismatchReport) {
	key := identity.dedupeKey()
	m.reportMu.Lock()
	if m.lastReportedKey == key {
		m.reportMu.Unlock()
		return
	}
	m.lastReportedKey = key
	m.reportMu.Unlock()
	callback := m.options.OnDefaultMismatch
	if callback == nil {
		return
	}
	func() {
		defer func() { _ = recover() }()
		callback(report)
	}()
}

// clearReported forgets the last reported identity so that re-activating a
// previously reported divergent release after a clean one is reported again.
func (m *Manager) clearReported() {
	m.reportMu.Lock()
	m.lastReportedKey = ""
	m.reportMu.Unlock()
}

func (r ReleaseIdentity) dedupeKey() string {
	return fmt.Sprintf("%s\x00%s\x00%d\x00%d\x00%s",
		r.namespace, r.name, r.version, r.activationRevision, r.digest)
}

type managedPrepared struct {
	manager    *Manager
	publish    func()
	abort      func()
	identity   ReleaseIdentity
	divergent  bool
	fieldCount int
	startup    bool
	changes    []FieldChange
	groups     func() (map[string]jsontext.Value, error)
}

func (p *managedPrepared) Commit() {
	// Publication is application-owned and must happen before readiness or
	// manager status becomes visible.
	p.publish()
	p.manager.mu.Lock()
	p.manager.applied = p.identity
	p.manager.divergent = p.divergent
	p.manager.ready = true
	p.manager.mu.Unlock()
	if !p.divergent {
		p.manager.clearReported()
	}
	p.manager.notifyApplied(p)
	p.manager.readyOnce.Do(func() { close(p.manager.readyCh) })
}

// ReleaseDivergence implements kmsclient.ReleaseDivergenceReporter so the
// applied acknowledgement carries divergence without any field detail.
func (p *managedPrepared) ReleaseDivergence() (bool, int) {
	return p.divergent, p.fieldCount
}

func (m *Manager) notifyApplied(p *managedPrepared) {
	callback := m.options.OnApplied
	if callback == nil {
		return
	}
	phase := PhaseRuntime
	if p.startup {
		phase = PhaseStartup
	}
	report := newAppliedReport(phase, p.identity, p.divergent, p.changes, p.groups)
	// Commit must be infallible for the loader; a callback panic must never
	// turn a published generation into a fatal loader failure.
	func() {
		defer func() { _ = recover() }()
		callback(report)
	}()
}

func (p *managedPrepared) Abort() { p.abort() }

func newAbort(abort func()) func() {
	if abort == nil {
		abort = func() {}
	}
	var once sync.Once
	return func() { once.Do(abort) }
}

type restartRequiredError struct {
	fields []string
}

func (e *restartRequiredError) Error() string {
	return "configstore: candidate changes restart-required fields: " + strings.Join(e.fields, ",")
}

// Status returns a redacted, concurrency-safe manager snapshot.
func (m *Manager) Status() Status {
	if m == nil {
		return Status{}
	}
	loaderStatus := m.loader.Status()
	m.mu.RLock()
	observed := m.observed
	if loaderStatus.ObservedVersion != observed.version || loaderStatus.ObservedRevision != observed.activationRevision {
		// Prefetch contract and resolution failures do not produce a resolved
		// snapshot. Preserve the safe version/revision observed by ReleaseLoader
		// while leaving unavailable schema/digest fields empty.
		observed = ReleaseIdentity{
			name:               m.options.Release,
			version:            loaderStatus.ObservedVersion,
			activationRevision: loaderStatus.ObservedRevision,
		}
	}
	status := Status{
		State:                 loaderStatus.State,
		Ready:                 m.ready,
		Observed:              observed,
		Applied:               m.applied,
		DefaultDivergent:      m.divergent,
		LastRejectionCategory: RejectionCategory(loaderStatus.LastFailureCategory),
		LastFailureAt:         loaderStatus.LastFailureAt,
		Reconnects:            loaderStatus.Reconnects,
	}
	m.mu.RUnlock()
	return status
}

// Stats returns bounded counters and a fresh rejection map.
func (m *Manager) Stats() Stats {
	if m == nil {
		return Stats{Rejected: make(map[RejectionCategory]uint64)}
	}
	loaderStats := m.loader.Stats()
	m.mu.RLock()
	stats := Stats{
		Candidates:                loaderStats.Candidates,
		Applied:                   loaderStats.Applied,
		Rejected:                  make(map[RejectionCategory]uint64, len(loaderStats.Rejected)),
		Reconnects:                loaderStats.Reconnects,
		DefaultDivergent:          m.divergent,
		AppliedReleaseVersion:     m.applied.version,
		AppliedActivationRevision: m.applied.activationRevision,
	}
	m.mu.RUnlock()
	for category, count := range loaderStats.Rejected {
		stats.Rejected[RejectionCategory(category)] = count
	}
	return stats
}

// Wait blocks until the background loader exits. Caller context cancellation
// after initial readiness is a normal shutdown and returns nil.
func (m *Manager) Wait() error {
	if m == nil {
		return errors.New("configstore: nil Manager")
	}
	<-m.done
	m.mu.RLock()
	err := m.waitErr
	m.mu.RUnlock()
	return err
}

File: sdk/typescript/src/configstore/manager.ts
import type { ClientReleaseLoaderOptions } from "../client.js";
import type { ReleaseLoader, ReleaseLoaderOptions } from "../releases/loader.js";
import type {
  PreparedRelease,
  ReleaseLoaderStatus,
  ReleaseRejectionCategory,
  ReleaseSnapshot,
} from "../releases/types.js";
import { cloneConfig } from "./clone.js";
import { type ContractEntry, createManifestValidator, validateContract } from "./contract.js";
import {
  AppliedReport,
  CandidateError,
  CandidateRejectionReport,
  candidateErrorPaths,
  DefaultMismatchReport,
  type FieldChange,
  type FieldDifference,
} from "./errors.js";
import { ReleaseIdentity } from "./snapshot.js";

export interface ManagedPreparedCandidate {
  /** Generated binding's atomic active-generation swap. Must be infallible. */
  readonly publish: () => undefined;
  /** Releases candidate-owned resources. Must be infallible and is called at most once. */
  readonly abort?: () => undefined;
  /** Complete non-secret comparison against source defaults. */
  readonly defaultDifferences?: readonly FieldDifference[];
  /** Generated canonical fields whose effective value/pin changed after startup. */
  readonly restartRequiredFields?: readonly string[];
  /**
   * Fields that differ from the previously applied generation; secret entries
   * must be path-only. Ignored for the initial generation.
   */
  readonly changed?: readonly FieldChange[];
  /** Canonical non-secret parameter group documents keyed by alias. Never secrets. */
  readonly groups?: Readonly<Record<string, string>>;
}

export type PrepareManagedCandidate = (
  snapshot: ReleaseSnapshot,
  signal: AbortSignal,
) => ManagedPreparedCandidate | Promise<ManagedPreparedCandidate>;

/**
 * Application-owned observers of a managed store. `onDefaultMismatch` is
 * required; the others are optional. Every callback runs synchronously on the
 * loader's candidate path and must not block; thrown errors are isolated and
 * returned promises are discarded. `consoleCallbacks` is a ready-made
 * implementation.
 */
export interface Callbacks {
  /**
   * Reports every candidate whose non-secret values differ from source
   * defaults, at startup and on every reload. The candidate is still applied;
   * the report is the signal to reconcile code and KMS. Must be synchronous.
   */
  readonly onDefaultMismatch: (report: DefaultMismatchReport) => void;
  /**
   * Fires after each generation is published, including the initial one,
   * carrying the fields that changed since the previously applied generation.
   */
  readonly onApplied?: (report: AppliedReport) => void;
  /** Optional, value-free local diagnostics. Callback failures are isolated. */
  readonly onCandidateRejected?: (report: CandidateRejectionReport) => void;
}

export interface ManagedConfigOptions
  extends Omit<
      ClientReleaseLoaderOptions,
      "name" | "namespace" | "clientName" | "validateManifest"
    >,
    Callbacks {
  /** Release name within the client's home or explicitly selected namespace. */
  readonly release: string;
  /** Optional env/app override. Omit to use the KmsClient home namespace. */
  readonly namespace?: string;
  /** Optional acknowledgement identity override. */
  readonly clientName?: string;
  readonly contract: readonly ContractEntry[];
}

/** Structural KmsClient subset; transport internals remain private. */
export interface ManagedReleaseClient {
  createReleaseLoader(options: ClientReleaseLoaderOptions): Promise<ReleaseLoader>;
}

export interface ManagedConfigStatus {
  readonly state: ReleaseLoaderStatus["state"];
  readonly ready: boolean;
  readonly observed: ReleaseIdentity;
  readonly applied: ReleaseIdentity;
  readonly defaultDivergent: boolean;
  readonly lastRejectionCategory?: ReleaseRejectionCategory;
  readonly lastFailureAt?: Date;
  readonly reconnects: bigint;
}

export interface ManagedConfigStats {
  readonly candidates: bigint;
  readonly applied: bigint;
  readonly rejected: Readonly<Record<string, bigint>>;
  readonly reconnects: bigint;
  readonly defaultDivergent: boolean;
  readonly appliedReleaseVersion: bigint;
  readonly appliedActivationRevision: bigint;
}

interface PolicyOptions extends Callbacks {
  readonly name: string;
}

interface Completion {
  readonly error?: unknown;
}

/**
 * Managed drift/restart policy around ReleaseLoader. Construct with
 * startManagedConfig so exact manifest validation runs before any resource
 * fetch and startup does not return before generated state is publishable.
 */
export class ManagedConfigManager {
  readonly #loader: ReleaseLoader;
  readonly #options: PolicyOptions;
  readonly #prepare: PrepareManagedCandidate;
  readonly #readySignal = deferred<void>();
  #completion: Promise<Completion> | undefined;
  #ready = false;
  #observed = new ReleaseIdentity();
  #applied = new ReleaseIdentity();
  #divergent = false;
  #lastReportedKey = "";
  #lastRejectedKey = "";

  /** @internal Use startManagedConfig. */
  constructor(loader: ReleaseLoader, options: PolicyOptions, prepare: PrepareManagedCandidate) {
    this.#loader = loader;
    this.#options = options;
    this.#prepare = prepare;
  }

  /** @internal Records safe unresolved identity before exact contract validation. */
  observeManifest(manifest: Parameters<ReturnType<typeof createManifestValidator>>[0]): void {
    this.#observed = ReleaseIdentity.from(manifest);
  }

  /** @internal Reports classified pre-resolution contract rejection locally. */
  rejectManifest(
    manifest: Parameters<ReturnType<typeof createManifestValidator>>[0],
    error: unknown,
  ): void {
    this.#notifyCandidateRejected(ReleaseIdentity.from(manifest), error);
  }

  /** @internal Starts the one owned loader run. */
  start(signal?: AbortSignal): void {
    if (this.#completion) throw new Error("configstore: manager is already started");
    const run = this.#loader.run((snapshot, candidateSignal) => {
      return this.#prepareSnapshot(snapshot, candidateSignal);
    }, signal);
    this.#completion = run.then(
      () => ({}),
      (error: unknown) => ({ error }),
    );
  }

  /** @internal Wait for the first atomic publication or the loader's startup failure. */
  async waitUntilReady(): Promise<void> {
    const completion = this.#completion;
    if (!completion) throw new Error("configstore: manager is not started");
    const outcome = await Promise.race([
      this.#readySignal.promise.then(() => ({ ready: true }) as const),
      completion.then((result) => ({ ready: false, result }) as const),
    ]);
    if (outcome.ready) return;
    throw (
      outcome.result.error ?? new Error("configstore: release loader stopped before publication")
    );
  }

  /** Request background loader shutdown. */
  stop(reason?: unknown): void {
    this.#loader.stop(reason);
  }

  /** Wait for loader termination. Cancellation after readiness is normal shutdown. */
  async wait(): Promise<void> {
    const completion = this.#completion;
    if (!completion) throw new Error("configstore: manager is not started");
    const { error } = await completion;
    if (error === undefined || (this.#ready && isAbortError(error))) return;
    throw error;
  }

  /** Redacted point-in-time lifecycle status. */
  status(): ManagedConfigStatus {
    const loaderStatus = this.#loader.status();
    let observed = this.#observed;
    if (
      observed.version !== loaderStatus.observedVersion ||
      observed.activationRevision !== loaderStatus.observedRevision
    ) {
      observed = new ReleaseIdentity({
        name: this.#options.name,
        version: loaderStatus.observedVersion,
        activationRevision: loaderStatus.observedRevision,
      });
    }
    return Object.freeze({
      state: loaderStatus.state,
      ready: this.#ready,
      observed,
      applied: this.#applied,
      defaultDivergent: this.#divergent,
      ...(loaderStatus.lastFailureCategory
        ? { lastRejectionCategory: loaderStatus.lastFailureCategory }
        : {}),
      ...(loaderStatus.lastFailureAt
        ? { lastFailureAt: new Date(loaderStatus.lastFailureAt) }
        : {}),
      reconnects: loaderStatus.reconnects,
    });
  }

  /** Fresh bounded metrics counters. */
  stats(): ManagedConfigStats {
    const loaderStats = this.#loader.stats();
    return Object.freeze({
      candidates: loaderStats.candidates,
      applied: loaderStats.applied,
      rejected: Object.freeze({ ...loaderStats.rejected }),
      reconnects: loaderStats.reconnects,
      defaultDivergent: this.#divergent,
      appliedReleaseVersion: this.#applied.version,
      appliedActivationRevision: this.#applied.activationRevision,
    });
  }

  async #prepareSnapshot(snapshot: ReleaseSnapshot, signal: AbortSignal): Promise<PreparedRelease> {
    const identity = ReleaseIdentity.from(snapshot);
    this.#observed = identity;
    const startup = !this.#ready;

    let candidate: ManagedPreparedCandidate;
    try {
      candidate = await this.#prepare(snapshot, signal);
    } catch (error) {
      this.#notifyCandidateRejected(identity, error);
      throw error;
    }
    const abort = once(candidate?.abort ?? (() => undefined));
    if (!candidate || typeof candidate.publish !== "function") {
      this.#abortOrInternal(abort, identity);
      const error = new CandidateError(
        "config_validation_failed",
        new Error("configstore: prepared candidate publish is required"),
      );
      this.#notifyCandidateRejected(identity, error);
      throw error;
    }
    if (signal.aborted) {
      this.#abortOrInternal(abort, identity);
      throw abortReason(signal);
    }

    let differences: FieldDifference[];
    let restartFields: string[];
    let changes: FieldChange[];
    let groups: Record<string, string>;
    try {
      differences = cloneDifferences(candidate.defaultDifferences ?? []);
      restartFields = cloneRestartFields(candidate.restartRequiredFields ?? []);
      // Always validate the binding's contract; the initial generation has no
      // previous generation, so its change list is discarded rather than reported.
      const candidateChanges = cloneChanges(candidate.changed ?? []);
      changes = startup ? [] : candidateChanges;
      groups = cloneGroups(candidate.groups ?? {});
    } catch (cause) {
      this.#abortOrInternal(abort, identity);
      const error = new CandidateError("internal", cause);
      this.#notifyCandidateRejected(identity, error);
      throw error;
    }
    if (!startup && restartFields.length > 0) {
      this.#abortOrInternal(abort, identity);
      const error = new CandidateError(
        "restart_required",
        new Error("configstore: candidate changes restart-required fields"),
        restartFields,
      );
      this.#notifyCandidateRejected(identity, error);
      throw error;
    }

    // Divergence from source defaults is reported, never refused: a process
    // must be able to restart onto whatever release is active. The report is
    // the operator's signal to reconcile code and KMS.
    const phase = startup ? "startup" : "runtime";
    const divergent = differences.length > 0;
    if (divergent) {
      const report = new DefaultMismatchReport(phase, "error", identity, differences);
      this.#reportOnce(identity, report);
    }

    return {
      commit: () => {
        const returned: unknown = candidate.publish();
        const contractError = synchronousCallbackError(
          "configstore: prepared candidate publish",
          returned,
        );
        if (contractError) throw contractError;
        this.#applied = identity;
        this.#divergent = divergent;
        this.#ready = true;
        // A clean generation forgets the last reported identity so rolling back
        // onto a previously reported divergent release is reported again.
        if (!divergent) this.#lastReportedKey = "";
        this.#readySignal.resolve();
        this.#notifyApplied(phase, identity, divergent, changes, groups);
        return undefined;
      },
      abort,
      releaseDivergence: () => ({ divergent, fieldCount: differences.length }),
    };
  }

  #abortOrInternal(abort: () => undefined, identity: ReleaseIdentity): void {
    try {
      const returned: unknown = abort();
      const contractError = synchronousCallbackError(
        "configstore: prepared candidate abort",
        returned,
      );
      if (contractError) throw contractError;
    } catch (cause) {
      const error = new CandidateError("internal", cause);
      this.#notifyCandidateRejected(identity, error);
      throw error;
    }
  }

  #notifyApplied(
    phase: "startup" | "runtime",
    identity: ReleaseIdentity,
    divergent: boolean,
    changes: readonly FieldChange[],
    groups: Readonly<Record<string, string>>,
  ): void {
    const callback = this.#options.onApplied;
    if (!callback) return;
    // commit must stay infallible for the loader; a callback failure must
    // never turn a published generation into a fatal loader failure.
    try {
      const report = new AppliedReport(phase, identity, divergent, changes, groups);
      const result = callUnknown(callback, report);
      discardPromise(result);
    } catch {
      // Diagnostics cannot affect an already published generation.
    }
  }

  #notifyCandidateRejected(identity: ReleaseIdentity, error: unknown): void {
    if (!(error instanceof CandidateError) || !this.#options.onCandidateRejected) return;
    const key = identity.dedupeKey();
    if (this.#lastRejectedKey === key) return;
    this.#lastRejectedKey = key;
    const report = new CandidateRejectionReport(
      error.category,
      identity,
      candidateErrorPaths(error),
    );
    try {
      const result = callUnknown(this.#options.onCandidateRejected, report);
      discardPromise(result);
    } catch {
      // Diagnostics cannot affect candidate admission or reconciliation.
    }
  }

  /**
   * Deliver a mismatch report at most once per release identity. The callback
   * is an observer: a throw or a returned promise is isolated and never
   * influences candidate admission, so a broken logger cannot refuse startup.
   */
  #reportOnce(identity: ReleaseIdentity, report: DefaultMismatchReport): void {
    const key = identity.dedupeKey();
    if (this.#lastReportedKey === key) return;
    this.#lastReportedKey = key;
    try {
      const result = callUnknown(this.#options.onDefaultMismatch, report);
      if (isPromiseLike(result)) discardPromise(result);
    } catch {
      // Isolated: observers never affect admission.
    }
  }
}

/**
 * Validate the generated contract, start ReleaseLoader, and resolve only after
 * the initial complete candidate has been atomically published.
 */
export async function startManagedConfig(
  client: ManagedReleaseClient,
  options: ManagedConfigOptions,
  prepare: PrepareManagedCandidate,
  signal?: AbortSignal,
): Promise<ManagedConfigManager> {
  if (!client || typeof client.createReleaseLoader !== "function") {
    throw new TypeError("configstore: KmsClient-compatible release client is required");
  }
  if (typeof prepare !== "function")
    throw new TypeError("configstore: prepare callback is required");
  if (typeof options?.onDefaultMismatch !== "function") {
    throw new TypeError("configstore: onDefaultMismatch callback is required");
  }
  if (options.onApplied !== undefined && typeof options.onApplied !== "function") {
    throw new TypeError("configstore: onApplied must be a function when provided");
  }
  const release = options.release?.trim();
  if (!release) throw new TypeError("configstore: release is required");
  const contract = validateContract(options.contract);
  let manager: ManagedConfigManager | undefined;
  const baseManifestValidator = createManifestValidator(contract);
  const validateManifest: NonNullable<ReleaseLoaderOptions["validateManifest"]> = (
    manifest,
    candidateSignal,
  ) => {
    manager?.observeManifest(manifest);
    try {
      return baseManifestValidator(manifest, candidateSignal);
    } catch (error) {
      manager?.rejectManifest(manifest, error);
      throw error;
    }
  };
  const loader = await client.createReleaseLoader(
    copyLoaderOptions(options, release, validateManifest),
  );
  manager = new ManagedConfigManager(
    loader,
    {
      name: release,
      onDefaultMismatch: options.onDefaultMismatch,
      ...(options.onApplied ? { onApplied: options.onApplied } : {}),
      ...(options.onCandidateRejected ? { onCandidateRejected: options.onCandidateRejected } : {}),
    },
    prepare,
  );
  manager.start(signal);
  await manager.waitUntilReady();
  return manager;
}

function copyLoaderOptions(
  options: ManagedConfigOptions,
  release: string,
  validateManifest: NonNullable<ReleaseLoaderOptions["validateManifest"]>,
): ClientReleaseLoaderOptions {
  return {
    name: release,
    ...(options.namespace ? { namespace: options.namespace } : {}),
    ...(options.clientName ? { clientName: options.clientName } : {}),
    ...(options.instanceId ? { instanceId: options.instanceId } : {}),
    ...(options.reconcileIntervalMs !== undefined
      ? { reconcileIntervalMs: options.reconcileIntervalMs }
      : {}),
    ...(options.maxConcurrentFetches !== undefined
      ? { maxConcurrentFetches: options.maxConcurrentFetches }
      : {}),
    ...(options.secretTokenProvider ? { secretTokenProvider: options.secretTokenProvider } : {}),
    validateManifest,
    ...(options.now ? { now: options.now } : {}),
    ...(options.random ? { random: options.random } : {}),
  };
}

function cloneDifferences(differences: readonly FieldDifference[]): FieldDifference[] {
  if (!Array.isArray(differences)) {
    throw new TypeError("configstore: default differences must be an array");
  }
  return differences.map((difference) => ({
    path: requireString(difference?.path, "default difference path"),
    expected: cloneConfig(difference.expected),
    actual: cloneConfig(difference.actual),
  }));
}

function cloneChanges(changes: readonly FieldChange[]): FieldChange[] {
  if (!Array.isArray(changes)) {
    throw new TypeError("configstore: changed fields must be an array");
  }
  return changes.map((change) => ({
    path: requireString(change?.path, "changed field path"),
    previous: cloneConfig(change.previous),
    current: cloneConfig(change.current),
  }));
}

function cloneGroups(groups: Readonly<Record<string, string>>): Record<string, string> {
  if (typeof groups !== "object" || groups === null || Array.isArray(groups)) {
    throw new TypeError("configstore: parameter groups must be a record of documents");
  }
  const result: Record<string, string> = Object.create(null) as Record<string, string>;
  for (const alias of Object.keys(groups)) {
    const descriptor = Object.getOwnPropertyDescriptor(groups, alias);
    if (!descriptor || !("value" in descriptor)) {
      throw new TypeError("configstore: parameter group documents must be data properties");
    }
    result[alias] = requireString(descriptor.value, "parameter group document");
  }
  return result;
}

function cloneRestartFields(fields: readonly string[]): string[] {
  if (!Array.isArray(fields)) {
    throw new TypeError("configstore: restart-required fields must be an array");
  }
  return fields.map((path) => requireString(path, "restart-required field path"));
}

function requireString(value: unknown, description: string): string {
  if (typeof value !== "string")
    throw new TypeError(`configstore: ${description} must be a string`);
  return value;
}

function once(callback: () => undefined): () => undefined {
  let called = false;
  return () => {
    if (called) return undefined;
    called = true;
    return callback();
  };
}

function synchronousCallbackError(name: string, returned: unknown): Error | undefined {
  if (returned === undefined) return undefined;
  void Promise.resolve(returned).catch(() => undefined);
  return new TypeError(`${name} must return undefined synchronously`);
}

function isAbortError(error: unknown): boolean {
  return (
    (error instanceof DOMException && error.name === "AbortError") ||
    (error instanceof Error && error.name === "AbortError")
  );
}

function abortReason(signal: AbortSignal): unknown {
  return signal.reason ?? new DOMException("Aborted", "AbortError");
}

function callUnknown<T>(callback: (value: T) => void, value: T): unknown {
  return (callback as (item: T) => unknown)(value);
}

function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
  return (
    (typeof value === "object" || typeof value === "function") &&
    value !== null &&
    "then" in value &&
    typeof value.then === "function"
  );
}

function discardPromise(value: unknown): void {
  if (isPromiseLike(value)) void Promise.resolve(value).catch(() => undefined);
}

function deferred<T>(): { readonly promise: Promise<T>; readonly resolve: (value: T) => void } {
  let resolve!: (value: T) => void;
  const promise = new Promise<T>((onResolve) => {
    resolve = onResolve;
  });
  return { promise, resolve };
}
